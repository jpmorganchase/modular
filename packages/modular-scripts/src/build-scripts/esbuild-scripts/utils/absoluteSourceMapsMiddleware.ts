import * as path from 'path';
import * as fs from 'fs-extra';
import type { Request, Response } from 'express';

type SourceMap = { sources: string[] };

/* This middleware remedies this issue - https://github.com/evanw/esbuild/pull/1234,
   where paths in sources that are relative to outDir break react-error-overlay.
   This is to be used before static.
*/

export function createAbsoluteSourceMapMiddleware(outDir: string) {
  return function sourceMapMiddleware(req: Request, res: Response): void {
    const filePath = path.join(outDir, req.path);
    void fs.readJson(filePath).then((sourceMapObject: SourceMap) => {
      // Make all paths absolute
      sourceMapObject.sources = sourceMapObject.sources.map((sourcePath) =>
        path.join(outDir, sourcePath),
      );
      res.json(sourceMapObject);
    });
  };
}
