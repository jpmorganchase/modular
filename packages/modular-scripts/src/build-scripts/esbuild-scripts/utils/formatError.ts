import * as path from 'path';
import { Message } from 'esbuild';
import chalk from 'chalk';
import * as fs from 'fs-extra';
import { codeFrameColumns } from '@babel/code-frame';

export async function formatError(
  error: Message,
  baseDir: string = process.cwd(),
): Promise<string> {
  if (error.location?.file) {
    const pathToFile = path.join(baseDir, error.location?.file);
    const source = await fs.readFile(pathToFile, {
      encoding: 'utf-8',
    });
    return `${chalk.red('Error:')}[${error.location.file}] ${error.text}
  
${codeFrameColumns(
  source,
  { start: { line: error.location.line, column: error.location.column } },
  {
    highlightCode: true,
  },
)}
    `;
  } else {
    return `${chalk.red('Error:')} ${error.text}\n`;
  }
}
