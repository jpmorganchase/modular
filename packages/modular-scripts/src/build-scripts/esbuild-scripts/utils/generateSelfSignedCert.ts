import fs from 'fs';
import path from 'path';
import { promisify } from 'util';
import selfsigned from 'selfsigned';
import rimraf from 'rimraf';
import * as logger from '../../../utils/logger';

const del = promisify(rimraf);

// This code is mostly duplicated from webpack-dev-server, with some typings added.
// To provide consistent https behaviour in esbuild mode, we use the same directories to store the generated cert.
// See: https://github.com/webpack/webpack-dev-server/blob/master/lib/Server.js

function findCacheDir() {
  const cwd = process.cwd();

  let dir: string | undefined = cwd;

  for (;;) {
    try {
      if (fs.statSync(path.join(dir, 'package.json')).isFile()) break;
    } catch (e) {}

    const parent = path.dirname(dir);

    if (dir === parent) {
      dir = undefined;
      break;
    }

    dir = parent;
  }

  if (!dir) {
    return path.resolve(cwd, '.cache/webpack-dev-server');
  } else if (process.versions.pnp === '1') {
    return path.resolve(dir, '.pnp/.cache/webpack-dev-server');
  } else if (process.versions.pnp === '3') {
    return path.resolve(dir, '.yarn/.cache/webpack-dev-server');
  }

  return path.resolve(dir, 'node_modules/.cache/webpack-dev-server');
}

export async function generateSelfSignedCert() {
  const certificateDir = findCacheDir();
  const certificatePath = path.join(certificateDir, 'server.pem');
  let certificateExists = fs.existsSync(certificatePath);

  if (certificateExists) {
    const certificateTtl = 1000 * 60 * 60 * 24;
    const certificateStat = fs.statSync(certificatePath);

    const now = new Date();

    // cert is more than 30 days old, kill it with fire
    if ((Number(now) - Number(certificateStat.ctime)) / certificateTtl > 30) {
      logger.log('SSL Certificate is more than 30 days old. Removing...');

      await del(certificatePath);

      certificateExists = false;
    }
  }

  if (!certificateExists) {
    logger.log('Generating SSL Certificate...');

    const attributes = [{ name: 'commonName', value: 'localhost' }];
    const pems = selfsigned.generate(attributes, {
      algorithm: 'sha256',
      days: 30,
      keySize: 2048,
      extensions: [
        {
          name: 'basicConstraints',
          cA: true,
        },
        {
          name: 'keyUsage',
          keyCertSign: true,
          digitalSignature: true,
          nonRepudiation: true,
          keyEncipherment: true,
          dataEncipherment: true,
        },
        {
          name: 'extKeyUsage',
          serverAuth: true,
          clientAuth: true,
          codeSigning: true,
          timeStamping: true,
        },
        {
          name: 'subjectAltName',
          altNames: [
            {
              // type 2 is DNS
              type: 2,
              value: 'localhost',
            },
            {
              type: 2,
              value: 'localhost.localdomain',
            },
            {
              type: 2,
              value: 'lvh.me',
            },
            {
              type: 2,
              value: '*.lvh.me',
            },
            {
              type: 2,
              value: '[::1]',
            },
            {
              // type 7 is IP
              type: 7,
              ip: '127.0.0.1',
            },
            {
              type: 7,
              ip: 'fe80::1',
            },
          ],
        },
      ],
    });

    fs.mkdirSync(certificateDir, { recursive: true });
    fs.writeFileSync(certificatePath, pems.private + pems.cert, {
      encoding: 'utf8',
    });
  }

  const fakeCert = fs.readFileSync(certificatePath);

  logger.log(`SSL certificate: ${certificatePath}`);

  return fakeCert;
}
