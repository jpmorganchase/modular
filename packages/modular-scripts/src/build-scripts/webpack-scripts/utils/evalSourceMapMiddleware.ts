import Server from 'webpack-dev-server';
import webpack from 'webpack';
import type {
  NextFunction,
  Request,
  Response,
  Stats,
} from 'webpack-dev-server';

function getBase64SourceMap(source: webpack.sources.Source) {
  const base64 = Buffer.from(JSON.stringify(source.map()), 'utf8').toString(
    'base64',
  );
  return `data:application/json;charset=utf-8;base64,${base64}`;
}

function getSourceById(stats: Stats, id: string) {
  const module = Array.from(stats.compilation.modules).find(
    (m) => stats.compilation.chunkGraph.getModuleId(m) === id,
  );
  return module ? module.originalSource() : null;
}

/*
 * Middleware responsible for retrieving a generated source
 * Receives a webpack internal url: "webpack-internal:///<module-id>"
 * Returns a generated source: "<source-text><sourceMappingURL><sourceURL>"
 *
 * Based on EvalSourceMapDevToolModuleTemplatePlugin.js
 */
export default function createEvalSourceMapMiddleware(server: Server) {
  return function handleWebpackInternalMiddleware(
    req: Request,
    res: Response,
    next: NextFunction,
  ) {
    if (req.url.startsWith('/__get-internal-source')) {
      const fileName = req.query.fileName as string;
      const match = fileName.match(/webpack-internal:\/\/\/(.+)/);
      const id = match ? match[1] : undefined;
      // Server.stats is private so we get around it this way
      const stats = server['stats'] as Stats | undefined;
      if (!id || !stats) {
        next();
      } else {
        const source = getSourceById(stats, id);
        if (source) {
          const base64SourceMap = getBase64SourceMap(source);
          const sourceMapURL = `//# sourceMappingURL=${base64SourceMap}`;

          const sourceURL = `//# sourceURL=webpack-internal:///${module.id}`;
          res.end(
            `${source.source() as string}\n${sourceMapURL}\n${sourceURL}`,
          );
        }
      }
    } else {
      next();
    }
  };
}
