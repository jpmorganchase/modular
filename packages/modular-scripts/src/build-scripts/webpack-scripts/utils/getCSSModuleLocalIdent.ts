import path from 'path';
import loaderUtils from 'loader-utils';

export default function getLocalIdent(
  context: { resourcePath: string; rootContext: string },
  localIdentName: unknown,
  localName: string,
  options: unknown,
) {
  // Use the filename or folder name, based on some uses the index.js / index.module.(css|scss|sass) project style
  const fileNameOrFolder = context.resourcePath.match(
    /index\.module\.(css|scss|sass)$/,
  )
    ? '[folder]'
    : '[name]';
  // Create a hash based on a the file location and class name. Will be unique across a project, and close to globally unique.
  const hash = loaderUtils.getHashDigest(
    Buffer.from(
      path.posix.relative(context.rootContext, context.resourcePath) +
        localName,
    ),
    'md5',
    'base64',
    5,
  );
  // Use loaderUtils to find the file or folder name
  const className = loaderUtils.interpolateName(
    context,
    fileNameOrFolder + '_' + localName + '__' + hash,
    options,
  );
  // Remove the .module that appears in every classname when based on the file and replace all "." with "_".
  return className.replace('.module_', '_').replace(/\./g, '_');
}
