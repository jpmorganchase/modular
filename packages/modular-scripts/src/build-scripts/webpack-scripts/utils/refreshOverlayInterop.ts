import { dismissRuntimeErrors, reportRuntimeError } from 'react-error-overlay';

module.exports = {
  clearRuntimeErrors: dismissRuntimeErrors,
  handleRuntimeError: reportRuntimeError,
};
