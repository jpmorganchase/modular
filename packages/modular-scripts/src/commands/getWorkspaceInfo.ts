import * as fs from 'fs-extra';
import * as path from 'path';

import { PackageType, ModularPackageJson } from '../utils/isModularType';
import { getAllWorkspaces } from '../utils/getAllWorkspaces';
import getModularRoot from '../utils/getModularRoot';

export interface WorkSpaceRecord {
  location: string;
  workspaceDependencies: string[];
  mismatchedWorkspaceDependencies: string[];
  type: PackageType;
  public: boolean;
}

type WorkspaceInfo = Record<string, WorkSpaceRecord>;

export async function getWorkspaceInfo(): Promise<WorkspaceInfo> {
  const workspace = getAllWorkspaces();
  const workspaceRoot = getModularRoot();

  const res: WorkspaceInfo = {};
  for (const [packageName, packageInfo] of Object.entries(workspace)) {
    const packageJson = (await fs.readJSON(
      path.join(workspaceRoot, packageInfo.location, 'package.json'),
    )) as ModularPackageJson;

    const type: PackageType = (packageJson.modular?.type ||
      'package') as PackageType;

    const modularPackageInfo = Object.assign(packageInfo, {
      type,
      public:
        type === ('package' as PackageType) ? !!packageJson.public : false,
    });

    res[packageName] = modularPackageInfo;
  }

  return res;
}
