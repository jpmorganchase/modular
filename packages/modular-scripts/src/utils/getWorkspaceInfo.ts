import * as fs from 'fs-extra';
import * as path from 'path';

import { ModularType, ModularPackageJson } from './isModularType';
import { getAllWorkspaces } from './getAllWorkspaces';
import getModularRoot from './getModularRoot';
import memoize from './memoize';

export interface WorkSpaceRecord {
  location: string;
  workspaceDependencies: string[];
  mismatchedWorkspaceDependencies: string[];
  type: ModularType;
  public: boolean;
}

type WorkspaceInfo = Record<string, WorkSpaceRecord | undefined>;

export async function getWorkspaceInfo(): Promise<WorkspaceInfo> {
  const workspace = getAllWorkspaces();
  const workspaceRoot = getModularRoot();

  const res: WorkspaceInfo = {};
  for (const [packageName, packageInfo] of Object.entries(workspace)) {
    const packageJson = (await fs.readJSON(
      path.join(workspaceRoot, packageInfo.location, 'package.json'),
    )) as ModularPackageJson;

    const type = packageJson.modular?.type || ('package' as ModularType);

    const modularPackageInfo = Object.assign(packageInfo, {
      type,
      public: !!packageJson.public,
    });

    res[packageName] = modularPackageInfo;
  }

  return res;
}

export default memoize(getWorkspaceInfo);
