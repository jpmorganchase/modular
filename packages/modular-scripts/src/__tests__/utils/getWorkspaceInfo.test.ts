import { getWorkspaceInfo } from '../../utils/getWorkspaceInfo';

test('getWorkspaceInfo', async () => {
  const workspace = await getWorkspaceInfo();
  Object.values(workspace).forEach((pkg) => {
    expect(pkg).toMatchSnapshot({ version: expect.any(String) as unknown });
  });
});
