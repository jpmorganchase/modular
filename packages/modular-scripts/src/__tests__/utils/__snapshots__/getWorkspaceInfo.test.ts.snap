// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getWorkspaceInfo 1`] = `
Object {
  "location": "packages/create-modular-react-app",
  "mismatchedWorkspaceDependencies": Array [],
  "public": true,
  "type": "package",
  "version": Any<String>,
  "workspaceDependencies": Array [],
}
`;

exports[`getWorkspaceInfo 2`] = `
Object {
  "location": "packages/dependency-resolver",
  "mismatchedWorkspaceDependencies": Array [],
  "public": true,
  "type": "package",
  "version": Any<String>,
  "workspaceDependencies": Array [],
}
`;

exports[`getWorkspaceInfo 3`] = `
Object {
  "location": "packages/eslint-config-modular-app",
  "mismatchedWorkspaceDependencies": Array [],
  "public": true,
  "type": "package",
  "version": Any<String>,
  "workspaceDependencies": Array [],
}
`;

exports[`getWorkspaceInfo 4`] = `
Object {
  "location": "packages/modular-scripts",
  "mismatchedWorkspaceDependencies": Array [],
  "public": true,
  "type": "package",
  "version": Any<String>,
  "workspaceDependencies": Array [],
}
`;

exports[`getWorkspaceInfo 5`] = `
Object {
  "location": "packages/modular-site",
  "mismatchedWorkspaceDependencies": Array [],
  "public": false,
  "type": "app",
  "version": Any<String>,
  "workspaceDependencies": Array [],
}
`;

exports[`getWorkspaceInfo 6`] = `
Object {
  "location": "packages/modular-template-app",
  "mismatchedWorkspaceDependencies": Array [],
  "public": true,
  "type": "template",
  "version": Any<String>,
  "workspaceDependencies": Array [],
}
`;

exports[`getWorkspaceInfo 7`] = `
Object {
  "location": "packages/modular-template-esm-view",
  "mismatchedWorkspaceDependencies": Array [],
  "public": true,
  "type": "template",
  "version": Any<String>,
  "workspaceDependencies": Array [],
}
`;

exports[`getWorkspaceInfo 8`] = `
Object {
  "location": "packages/modular-template-node-env-app",
  "mismatchedWorkspaceDependencies": Array [],
  "public": false,
  "type": "template",
  "version": Any<String>,
  "workspaceDependencies": Array [],
}
`;

exports[`getWorkspaceInfo 9`] = `
Object {
  "location": "packages/modular-template-package",
  "mismatchedWorkspaceDependencies": Array [],
  "public": true,
  "type": "template",
  "version": Any<String>,
  "workspaceDependencies": Array [],
}
`;

exports[`getWorkspaceInfo 10`] = `
Object {
  "location": "packages/modular-template-view",
  "mismatchedWorkspaceDependencies": Array [],
  "public": true,
  "type": "template",
  "version": Any<String>,
  "workspaceDependencies": Array [],
}
`;

exports[`getWorkspaceInfo 11`] = `
Object {
  "location": "packages/modular-views.macro",
  "mismatchedWorkspaceDependencies": Array [],
  "public": true,
  "type": "package",
  "version": Any<String>,
  "workspaceDependencies": Array [],
}
`;

exports[`getWorkspaceInfo 12`] = `
Object {
  "location": "packages/tree-view-for-tests",
  "mismatchedWorkspaceDependencies": Array [],
  "public": false,
  "type": "package",
  "version": Any<String>,
  "workspaceDependencies": Array [],
}
`;
